src/txn/txn_rollback_nondurable.c

```diff
@@ -0,0 +1,319 @@
+		    "requires a timestamp greater than the pinned timestamp");
+
+	/*
+	 * Help the user - see if they have any active transactions. I'd
+	 * like to check the transaction running flag, but that would
+	 * require peeking into all open sessions, which isn't really
+	 * kosher.
+	 */
+	WT_ORDERED_READ(session_cnt, S2C(session)->session_cnt);
+	for (i = 0, s = txn_global->states; i < session_cnt; i++, s++) {
+		if (s->id != WT_TXN_NONE || s->pinned_id != WT_TXN_NONE)
+			WT_RET_MSG(session, EINVAL,
+			    "rollback_nondurable_commits not supported with "
+			    "active transactions");
+	}
+
+	return (0);
+}
+
+/*
+ * __txn_abort_newer_update --
+ *	Review an update chain
+ */
+static void
+__txn_abort_newer_update(WT_SESSION_IMPL *session,
+    WT_UPDATE *upd, wt_timestamp_t rollback_timestamp)
+{
+	bool aborted_one = false;
+	WT_UPDATE *next_upd;
+
+	for (next_upd = upd; next_upd != NULL; next_upd = next_upd->next) {
+		/* Only updates with timestamps will be aborted */
+		if (!__wt_timestamp_iszero(next_upd->timestamp) &&
+		    __wt_timestamp_cmp(
+		    rollback_timestamp, next_upd->timestamp) < 0) {
+			next_upd->txnid = WT_TXN_ABORTED;
+			/* TODO: should we clear the timestamp? */
+
+			/*
+			* If any updates are aborted, all newer updates
+			* better be aborted as well.
+			*/
+			if (!aborted_one)
+				WT_ASSERT(session,
+				    !aborted_one || upd == next_upd);
+			aborted_one = true;
+		}
+	}
+}
+
+/*
+ * __txn_abort_newer_row_skip --
+ *	Apply the update abort check to each entry in an insert skip list
+ */
+static void
+__txn_abort_newer_row_skip(WT_SESSION_IMPL *session,
+    WT_INSERT_HEAD *head, wt_timestamp_t rollback_timestamp)
+{
+	WT_INSERT *ins;
+
+	WT_SKIP_FOREACH(ins, head)
+		__txn_abort_newer_update(session, ins->upd, rollback_timestamp);
+}
+
+/*
+ * __txn_abort_newer_row_leaf --
+ *	Abort updates on a row leaf page with timestamps too new.
+ */
+static void
+__txn_abort_newer_row_leaf(
+    WT_SESSION_IMPL *session, WT_PAGE *page, wt_timestamp_t rollback_timestamp)
+{
+	WT_INSERT_HEAD *insert;
+	WT_ROW *rip;
+	WT_UPDATE *upd;
+	uint32_t i;
+
+	/*
+	 * Review the insert list for keys before the first entry on the disk
+	 * page.
+	 */
+	if ((insert = WT_ROW_INSERT_SMALLEST(page)) != NULL)
+		__txn_abort_newer_row_skip(
+		    session, insert, rollback_timestamp);
+
+	/*
+	 * Review updates that belong to keys that are on the disk image,
+	 * as well as for keys inserted since the page was read from disk.
+	 */
+	WT_ROW_FOREACH(page, rip, i) {
+		if ((upd = WT_ROW_UPDATE(page, rip)) != NULL)
+			__txn_abort_newer_update(
+			    session, upd, rollback_timestamp);
+
+		if ((insert = WT_ROW_INSERT(page, rip)) != NULL)
+			__txn_abort_newer_row_skip(
+			    session, insert, rollback_timestamp);
+	}
+}
+
+/*
+ * __txn_abort_newer_fast_deleted --
+ *	Fixup deleted refs so they are left in the correct state
+ */
+static void
+__txn_abort_newer_fast_deleted(WT_REF *ref, wt_timestamp_t rollback_timestamp)
```

Comment:
> What is the page_del timestamp? Why does it have one, as well as all the updates in the structure having one?

The page-del timestamp is the txn ID that deleted the page, I don't know what you mean by "all the updates in the structure having one".

> Why does WT_DELETED_REF hold an array of WT_UPDATE structures? It doesn't seem to be an array of WT_UPDATE chains, so why not just use the linked list approach of walking through WT_UPDATE.next.

IIRC, we create those update structures when instantiating a page that's been "fast deleted", and the fast delete hasn't yet committed or aborted. The idea is a reader wants a fast delete page, and so we build a real page, and pretend it was in memory all along, and all of the items on the page were deleted individually by the fast delete transaction. We create the individual WT_UPDATE structures on the page we're instantiating, but keep a list of them in case the fast delete transaction aborts, because aborting a transaction requires a walk of the transaction's updates, setting their state to "abort". Ideally, we'd somehow incorporate that list (of updates to abort), in the transaction's list of updates to abort, but that would mean the thread instantiating the page has to write to another transaction's list of things to abort, and that list isn't locked in any way. We don't want to lock that list, so we stuck this list of "things to abort" in the page-deleted structure instead.

We could add a locked field to the transaction, which is the list of "fast delete" updates that need to be included in this transaction's abort, that might be a lot simpler (and would shrink the `WT_REF`).

Let's talk this over, I'm happy to make a change like that separately in develop, if it would simplify your life.